Dapper/DefaultTypeMap.cs

```diff
@@ -77,6 +77,27 @@ internal static List<FieldInfo> GetSettableFields(Type t)
         /// <param name="types">DataReader column types</param>
         /// <returns>Matching constructor or default one</returns>
         public ConstructorInfo FindConstructor(string[] names, Type[] types)
+        {
+            return FindConstructorByColumns(names, types) // first search constructor using the same position of names
+                ?? FindConstructorBySortedNames(names, types); // if not found, search by names sorted (F# anonymous records)
+        }
+
+        private ConstructorInfo FindConstructorBySortedNames(string[] names, Type[] types)

```

Comment:
isn't most of this method just: `Array.Sort(names, types);` ? that would be a **lot** more efficient, if so

Reply:
I tried to understand how to use `Array.Sort(names, types)`, but dunno how to use it here, because `names` is not the keys for `types` array.

`names` contains the ctor arg name, `types` is the ctor arg type.
So `names[0]` is the name of the first arg, and `types[0]` is it's expected type

That's why i do

```csharp
var byName = names.Zip(types, (name, type) => new { name, type })
```

to pair each name with his type, so i can order

```
`.OrderBy(x => x.name, StringComparer.OrdinalIgnoreCase`
```

these maintaining the sync.

i'll later split it two array again

```csharp
var namesByName = byName.Select(x => x.name).ToArray();
var typesByName = byName.Select(x => x.type).ToArray();
```

because i want to reuse the existing implementation who find the constructor

```
return FindConstructorByColumns(namesByName, typesByName);
```

Reply:
I can maybe use `Array.Sort` but i need anyway to:

- create the paired array (to maintain sync)
- sort by ordinal the `names` array into `namesSorted`
- `Array.Sort` the paired array with the `namesSorted` array and with a custom comparer (to extract the name pair)

i'll also need to use a tuple instead of anynoumn type, if i need a custom comparer i think.

So will be two sort (instead of just one) and the same number of temp array instantiations.

Reply:
The only enhancement i can think of is to replace

```csharp
var namesByName = byName.Select(x => x.name).ToArray();
var typesByName = byName.Select(x => x.type).ToArray();
```

with

```csharp
string[] namesByName = new string[ byName.Length ];
Type[] typesByName = new Type[ byName.Length ];

for (int i = 0; i < byName.Length; ++i) {
   namesByName[i] = byName[i].name;
   typesByName[i] = byName[i].type;
}
```

who should be i think a bit faster, but less readable.

what do you think @mgravell ?